docs/reusing-typecheck-results.md

```diff
@@ -0,0 +1,317 @@
+
+**2) Create the universal serialization + deserialization mechanism right away and use it for everything**
+
+This can take inspiration from existing mechanisms and random libraries on the internet. Here we'll be able to unlock the full potential of data caching, yet it will also mean a bigger investment in the beginning. Also, the caching procedure would be more uniform here - we can probably store whatever we can per each file, in the same format. 
+
+### Cache invalidation
+
+It's clear that we should be as defensive as possible here, cancelling further information reuse when suspecting that it might not match the fresh recompilation info.
+
+**Checking source files**
+
+To understand the source file is the same we can:
+1) Check the timestamp
+2) Hash the content and compare the hashes
+
+The former might lead to problems with things like time zone changes or manual file metadata modification. The latter might lead to problems in case of hash collisions. All of those scenarios are probably very unlikely, we might just look at what MSBuild does.
+
+**Checking auxiliary information**
+
+Besides tracking the source files, we should track:
+- compiler version
+- compiler flags 
+- ?
+
+**Checking dependencies**
+
+This is related to the minimization of retypecheck in slightly modified projects. If only one file is changed, how do we know that nothing depends on that?
+
+Graph-based typechecking mechanism should help here, however this also means we might need to store some additional data to keep track of things.
+
+### Data propagation
+
+Since the compilation steps we want to omit happen in different phases of compilation, I see 2 approaches to data propagation:
+
+**1) Cache and restore data in each phase.**
+
+For example, independently restore typecheck data and optimization data just before doing the respective jobs. This would lead to redundant checks (we don't need to reoptimize if we don't retypecheck), hence some code clutter, but would keep the general code structure same. We'll also have to leak the data a bit (invalidate optimization cache if we invalidate typecheck cache).
+
+**2) Restore at the beginning, cache at the end.**
+
+This would mean doing all the data restoring either in `main1` or even in a new phase dedicated to just that. Then the information would be gradually collected and propagated to the `main6`. This will avoid unnecessary checks but will also mean modifying the `Args` data and hence significantly intervening into the current compiler logic. 
+
+## Testing and benchmarking
+
+We should have a feature flag for this: applying it to current typechecking tests shouldn't make any difference in results. Unit tests should test that restored cached results + reusing them should be equivalent to fresh typecheck results.
```

Comment:
Dedicated compiler switch, not a feature flag.

Reply:
Yep, fixed.